core/src/main/java/org/apache/iceberg/OverwriteData.java

```diff
@@ -88,6 +119,42 @@ public OverwriteFiles validateAddedFiles() {
       }
     }
 
+    if (conflictDetectionFilter != null) {
+      PartitionSpec spec = writeSpec();
+      Expression inclusiveExpr = Projections.inclusive(spec).project(conflictDetectionFilter);
+      Evaluator inclusive = new Evaluator(spec.partitionType(), inclusiveExpr);
+
+      InclusiveMetricsEvaluator metrics = new InclusiveMetricsEvaluator(base.schema(), conflictDetectionFilter);
+
+      List<DataFile> newFiles = collectNewFiles(base);
+      for (DataFile newFile : newFiles) {
+        ValidationException.check(
+            !inclusive.eval(newFile.partition()) || !metrics.eval(newFile),
+            "A conflicting file was appended that matches filter '%s': %s",
+            conflictDetectionFilter, newFile.path());
+      }
+    }
+
     return super.apply(base);
   }
+
+  private List<DataFile> collectNewFiles(TableMetadata meta) {
+    List<DataFile> newFiles = new ArrayList<>();
+
+    Long currentSnapshotId = meta.currentSnapshot() == null ? null : meta.currentSnapshot().snapshotId();
+    while (currentSnapshotId != null && !currentSnapshotId.equals(readSnapshotId)) {
```

Comment:
If `readSnapshotId` is null, then this use the _entire_ table history? Or am I missing where `readSnapshotId` is defaulted to the current snapshot ID when the operation starts?

Reply:
This part is only invoked if we call `validateNoConflictingAppends`, which always accepts a valid `readSnapshotId` and an expression for detecting conflicts. If `validateNoConflictingAppends` is not called, there will be no validation and the behavior would be as it was before this change. 

We will call `validateNoConflictingAppends(null, expr)` only if we start with an empty table, meaning that we have to scan the entire history.

We can consider setting `readSnapshotId` and `conflictDetectionFilter` independently, but I am not sure whether it gives us any benefits. Also, it will lead to correctness errors if we won't set `readSnapshotId` in transactions correctly.